MINIFICPP-1736 Add MiNiFi agent C2 operation capabilities to agent manifest

lordgamez opened the pull request:
Added MiNiFi supported C2 operation data to the agent manifest in the following format: 

```
{
  ..
  "agentInfo": {
      "agentManifest": {
          ...
          "supportedOperations": [
              {
                  "properties": {
                      <OPERAND1>: {},
                      <OPERAND2>: {
                          <METADATA_KEY>: [
                            <METADATA_VALUE_OBJECTS>
                          ]
                      }
                  },
                  "type": <OPERATION>
              }
          ],
          ...
      },
      ...
  }
}
```

https://issues.apache.org/jira/browse/MINIFICPP-1736

------------------------------------------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

martinzink commented on libminifi/src/FlowController.cpp:
```diff
@@ -431,8 +431,9 @@ int16_t FlowController::clearConnection(const std::string &connection) {
-std::shared_ptr<state::response::ResponseNode> FlowController::getAgentManifest() const {
+std::shared_ptr<state::response::ResponseNode> FlowController::getAgentManifest() {
   auto agentInfo = std::make_shared<state::response::AgentInformation>("agentInfo");
+  agentInfo->setUpdatePolicyController(std::static_pointer_cast<controllers::UpdatePolicyControllerService>(getControllerService(c2::C2Agent::UPDATE_NAME)));
```
Nonconst getters could  lead to some bad practices in my opinion.

Since we only use this UpdatePolicyControllerService to call const functions (namely canUpdate()) I think these update_policy_controller_ members could be made into a `shared_ptr<const controllers::UpdatePolicyControllerService>`
And if we were to create a bunch of boilerplate to enable getting a shared_ptr<const ControllerServiceNode> from a const FlowController, then we could keep getAgentManifest as a const function.

lordgamez replied:
Good point, I was a bit hesitant changing it because it was very low core level stuff and I thought it would be convoluted, but after trying it, it seems these controller provider getters could be made const quite easily. Updated it in 5e37b8b76b97c6e04ee145c80293b5557348ba66

lordgamez commented on extensions/http-curl/tests/C2NullConfiguration.cpp:
```diff
@@ -72,15 +72,15 @@ class VerifyC2Server : public HTTPIntegrationBase {
-    configuration->set("c2.rest.listener.port", "null");
-    configuration->set("c2.agent.heartbeat.period", "null");
-    configuration->set("c2.rest.listener.heartbeat.rooturi", "null");
+    configuration->set(org::apache::nifi::minifi::Configuration::nifi_c2_enable, "true");
+    configuration->set(org::apache::nifi::minifi::Configuration::nifi_c2_agent_class, "test");
+    configuration->set(org::apache::nifi::minifi::Configuration::nifi_c2_agent_protocol_class, "RESTSender");
+    configuration->set(org::apache::nifi::minifi::Configuration::nifi_c2_rest_url, "");
+    configuration->set(org::apache::nifi::minifi::Configuration::nifi_c2_rest_url_ack, "");
+    configuration->set(org::apache::nifi::minifi::Configuration::nifi_c2_agent_heartbeat_reporter_classes, "null");
+    configuration->set(org::apache::nifi::minifi::Configuration::nifi_c2_rest_listener_port, "null");
+    configuration->set(org::apache::nifi::minifi::Configuration::nifi_c2_agent_heartbeat_period, "null");
+    configuration->set(org::apache::nifi::minifi::Configuration::nifi_c2_rest_listener_heartbeat_rooturi, "null");
```
Do you mean moving the VerifyC2Server class under minifi namespace or something else?

lordgamez replied:
I agree, I was just curious as it is not common currently, updated in 530a1d11cce113c61dbbee173aa515f7baf6b716

lordgamez commented on extensions/coap/controllerservice/CoapConnector.cpp:
```diff
@@ -67,7 +67,7 @@ void CoapConnectorService::onEnable() {
   } else {
     // this is the case where we aren't being used in the context of a single controller service.
-    if (configuration_->get("nifi.c2.agent.coap.host", host_) && configuration_->get("nifi.c2.agent.coap.port", port_str)) {
+    if (configuration_->get(minifi::Configuration::nifi_c2_agent_coap_host, host_) && configuration_->get(minifi::Configuration::nifi_c2_agent_coap_port, port_str)) {
```
Update in 530a1d11cce113c61dbbee173aa515f7baf6b716

lordgamez commented on extensions/civetweb/protocols/RESTReceiver.cpp:
```diff
@@ -51,8 +52,8 @@ void RESTReceiver::initialize(core::controller::ControllerServiceProvider* contr
-    configuration_->get("nifi.c2.rest.listener.port", "c2.rest.listener.port", listeningPort);
-    configuration_->get("nifi.c2.rest.listener.cacert", "c2.rest.listener.cacert", caCert);
+    configuration_->get(minifi::Configuration::nifi_c2_rest_listener_port, "c2.rest.listener.port", listeningPort);
+    configuration_->get(minifi::Configuration::nifi_c2_rest_listener_cacert, "c2.rest.listener.cacert", caCert);
```
Update in 530a1d11cce113c61dbbee173aa515f7baf6b716

lordgamez commented on encrypt-config/tests/ConfigFileTests.cpp:
```diff
@@ -112,21 +113,21 @@ TEST_CASE("ConfigFile can parse a simple config file", "[encrypt-config][constru
-  REQUIRE(!test_file.hasValue("nifi.remote.input.secure"));  // commented out
+  REQUIRE(test_file.hasValue(org::apache::nifi::minifi::Configuration::nifi_version));
+  REQUIRE(test_file.hasValue(org::apache::nifi::minifi::Configuration::nifi_c2_flow_id));  // present but blank
+  REQUIRE(!test_file.hasValue(org::apache::nifi::minifi::Configuration::nifi_remote_input_secure));  // commented out
```
Update in 530a1d11cce113c61dbbee173aa515f7baf6b716

lordgamez commented on encrypt-config/ConfigFile.cpp:
```diff
@@ -23,11 +23,12 @@
-constexpr const char* ADDITIONAL_SENSITIVE_PROPS_PROPERTY_NAME = "nifi.sensitive.props.additional.keys";
+constexpr std::array<const char*, 2> DEFAULT_SENSITIVE_PROPERTIES{org::apache::nifi::minifi::Configuration::nifi_security_client_pass_phrase,
+                                                                  org::apache::nifi::minifi::Configuration::nifi_rest_api_password};
+constexpr const char* ADDITIONAL_SENSITIVE_PROPS_PROPERTY_NAME = org::apache::nifi::minifi::Configuration::nifi_sensitive_props_additional_keys;
```
Update in 530a1d11cce113c61dbbee173aa515f7baf6b716

lordgamez replied:
Updated in 92ccbbbf2986de81653b471b0d0043ab988cc111

szaszm commented on libminifi/test/unit/SocketTests.cpp:
```diff
@@ -221,7 +222,7 @@ TEST_CASE("TestTLSContextCreation", "[TestSocket8]") {
 TEST_CASE("TestTLSContextCreation2", "[TestSocket9]") {
   std::shared_ptr<minifi::Configure> configure = std::make_shared<minifi::Configure>();
-  configure->set("nifi.remote.input.secure", "false");
+  configure->set(org::apache::nifi::minifi::Configuration::nifi_remote_input_secure, "false");
```
There is already a minifi namespace alias, use it!

lordgamez replied:
Updated in 107193764914b719074c45b1ce4b978c3d1292f4

lordgamez commented on libminifi/include/properties/Configuration.h:
```diff
@@ -99,6 +139,30 @@ class Configuration : public Properties {
+  static constexpr const char *nifi_log_compression_cached_log_max_size = "nifi.log.compression.cached.log.max.size";
+  static constexpr const char *nifi_log_compression_compressed_log_max_size = "nifi.log.compression.compressed.log.max.size";
+
+  MINIFIAPI static const std::vector<core::ConfigurationProperty> CONFIGURATION_PROPERTIES;
```
The reason the vector was introduced to be able to iterate through all the available properties to be added to the agent manifest.

lordgamez commented on libminifi/src/c2/C2Agent.cpp:
```diff
@@ -334,44 +334,7 @@ struct C2DebugBundleError : public C2TransferError {
-        logger_->log_debug("Clearing unknown %s", resp.name);
-      }
-
+      handle_clear(resp);
```
The definition of the function changed in both PRs in different ways, so it will definitely conflict either way. I would keep it this way for now and would resolve the conflict when the other PR is merged.

lordgamez commented on encrypt-config/tests/ConfigFileEncryptorTests.cpp:
```diff
@@ -76,13 +77,13 @@ TEST_CASE("ConfigFileEncryptor can encrypt the sensitive properties", "[encrypt-
   SECTION("default properties") {
     ConfigFile test_file{std::ifstream{"resources/minifi.properties"}};
-    std::string original_password = test_file.getValue("nifi.rest.api.password").value();
+    std::string original_password = test_file.getValue(org::apache::nifi::minifi::Configuration::nifi_rest_api_password).value();
```
Updated in 107193764914b719074c45b1ce4b978c3d1292f4

lordgamez commented on encrypt-config/ConfigFile.cpp:
```diff
@@ -23,11 +23,10 @@
-constexpr std::array<const char*, 2> DEFAULT_SENSITIVE_PROPERTIES{"nifi.security.client.pass.phrase",
-                                                                  "nifi.rest.api.password"};
-constexpr const char* ADDITIONAL_SENSITIVE_PROPS_PROPERTY_NAME = "nifi.sensitive.props.additional.keys";
+
```
Updated in 107193764914b719074c45b1ce4b978c3d1292f4

lordgamez commented on libminifi/include/core/Property.h:
```diff
@@ -467,6 +468,16 @@ class ConstrainedProperty : public std::enable_shared_from_this<ConstrainedPrope
+  }
+
+  std::string_view name;
+  gsl::not_null<std::shared_ptr<PropertyValidator>> validator;
```
Updated in 107193764914b719074c45b1ce4b978c3d1292f4

lordgamez commented on libminifi/src/Configure.cpp:
```diff
@@ -68,15 +69,15 @@ bool Configure::isEncrypted(const std::string& key) const {
 std::optional<std::string> Configure::getAgentClass() const {
   std::string agent_class;
-  if (get("nifi.c2.agent.class", "c2.agent.class", agent_class) && !agent_class.empty()) {
+  if (get(minifi::Configuration::nifi_c2_agent_class, "c2.agent.class", agent_class) && !agent_class.empty()) {
```
Updated in 107193764914b719074c45b1ce4b978c3d1292f4

lordgamez commented on libminifi/src/Configure.cpp:
```diff
@@ -68,15 +69,15 @@ bool Configure::isEncrypted(const std::string& key) const {
 std::string Configure::getAgentIdentifier() const {
   std::string agent_id;
-  if (!get("nifi.c2.agent.identifier", "c2.agent.identifier", agent_id) || agent_id.empty()) {
+  if (!get(minifi::Configuration::nifi_c2_agent_identifier, "c2.agent.identifier", agent_id) || agent_id.empty()) {
```
Updated in 107193764914b719074c45b1ce4b978c3d1292f4

lordgamez commented on libminifi/src/c2/C2Agent.cpp:
```diff
@@ -138,7 +138,7 @@ void C2Agent::configure(const std::shared_ptr<Configure> &configure, bool reconf
 
   if (!reconfigure) {
-    if (!configure->get("nifi.c2.agent.protocol.class", "c2.agent.protocol.class", clazz)) {
+    if (!configure->get(minifi::Configuration::nifi_c2_agent_protocol_class, "c2.agent.protocol.class", clazz)) {
```
Updated in 107193764914b719074c45b1ce4b978c3d1292f4

lordgamez commented on libminifi/src/c2/C2Agent.cpp:
```diff
@@ -164,7 +164,7 @@ void C2Agent::configure(const std::shared_ptr<Configure> &configure, bool reconf
   }
 
-  if (configure->get("nifi.c2.agent.heartbeat.period", "c2.agent.heartbeat.period", heartbeat_period)) {
+  if (configure->get(minifi::Configuration::nifi_c2_agent_heartbeat_period, "c2.agent.heartbeat.period", heartbeat_period)) {
```
Updated in 107193764914b719074c45b1ce4b978c3d1292f4

lordgamez commented on libminifi/src/c2/C2Agent.cpp:
```diff
@@ -181,7 +181,7 @@ void C2Agent::configure(const std::shared_ptr<Configure> &configure, bool reconf
 
   std::string heartbeat_reporters;
-  if (configure->get("nifi.c2.agent.heartbeat.reporter.classes", "c2.agent.heartbeat.reporter.classes", heartbeat_reporters)) {
+  if (configure->get(minifi::Configuration::nifi_c2_agent_heartbeat_reporter_classes, "c2.agent.heartbeat.reporter.classes", heartbeat_reporters)) {
```
Updated in 107193764914b719074c45b1ce4b978c3d1292f4

lordgamez commented:
The JSON structure of the supported operations was simplified in 9a29abe81ce6b918dcb1b9f8924117dc19a2be59, with some additional bugfixes/new features of the serialization:
- removed merging of parent-child nodes with the same name in `HeartbeatJsonSerializer.cpp` that would cause a bug of losing non-collapsible JSON nodes
- added `keep_empty` flag to Value nodes signaling to not drop empty JSON objects or arrays while serializing

lordgamez commented on C2.md:
```diff
@@ -148,22 +148,22 @@ configuration produces the following JSON:
 
 ### Protocols
 
-The default protocol is a RESTFul service; however, there is an MQTT protocol with a translation to use the 
```
I'm not sure how it works at the moment so it has to be checked I would leave the documentation as is for now.

lordgamez commented on extensions/http-curl/tests/HTTPHandlers.h:
```diff
@@ -385,15 +387,20 @@ std::string readPayload(struct mg_connection *conn) {
-              "text/plain\r\nContent-Length: %lu\r\nConnection: close\r\n\r\n",
-              resp.length());
-    mg_printf(conn, "%s", resp.c_str());
+  virtual void handleHeartbeat(const rapidjson::Document& root, struct mg_connection *) {
```
These methods are defined here first so they cannot be overriden.

lordgamez commented on extensions/http-curl/tests/HTTPHandlers.h:
```diff
@@ -446,13 +453,122 @@ class HeartbeatHandler : public ServerAwareHandler {
+  void verifyMetadata(const rapidjson::Value& operation_node, const std::unordered_map<std::string, Metadata>& operand_with_metadata) {
+    std::unordered_map<std::string, Metadata> operand_with_metadata_found;
+    const auto& properties_node = operation_node["properties"];
+    for (auto prop_it = properties_node.MemberBegin(); prop_it < properties_node.MemberEnd(); ++prop_it) {
```
It must have been a mistake as the `properties` node was previously an array and we were iterating through it using its `size()` method, so it should have been changed.

lordgamez replied:
Updated in b4c24f1f707c06aa35646e3881e0fbd9d5502e9b

lordgamez commented on libminifi/include/core/Property.h:
```diff
@@ -467,6 +468,16 @@ class ConstrainedProperty : public std::enable_shared_from_this<ConstrainedPrope
 
+struct ConfigurationProperty {
+  explicit ConfigurationProperty(std::string_view name_, gsl::not_null<PropertyValidator*> validator_ = gsl::make_not_null(StandardValidators::get().VALID_VALIDATOR.get()))
+    : name(name_),
+      validator(validator_) {
```
According to the [google style guide](https://google.github.io/styleguide/cppguide.html#Variable_Names) in case of structs we shouldn't use `_` suffixes for members, I used them on the local variables for distinguishing them instead.

fgerlits replied:
The style guide says not to use `_` for parameters, either, so we are deviating from it in either case.  I think using `_` for the struct members is the smaller deviation, and fits the rest of our code more.  But if you like it better this way, that's fine, too.

lordgamez replied:
Removed the suffixes altogether as it is not needed in the initializer list.

lordgamez commented on C2.md:
```diff
@@ -67,30 +67,30 @@ be requested via C2 DESCRIBE manifest command.
 	# nifi.c2.agent.protocol.class=CoapProtocol
-	
+
 	# control c2 heartbeat interval in millisecocnds
```
Updated in b4c24f1f707c06aa35646e3881e0fbd9d5502e9b

lordgamez commented on libminifi/include/core/state/nodes/SupportedOperations.h:
```diff
@@ -0,0 +1,69 @@
+
+  std::string getName() const override;
+  std::vector<SerializedResponseNode> serialize() override;
+  void setStateMonitor(const std::shared_ptr<state::StateMonitor> &monitor) {
```
Updated in b4c24f1f707c06aa35646e3881e0fbd9d5502e9b

lordgamez commented on libminifi/include/core/state/nodes/SupportedOperations.h:
```diff
@@ -0,0 +1,69 @@
+    monitor_ = monitor;
+  }
+
+  void setUpdatePolicyController(const std::shared_ptr<const controllers::UpdatePolicyControllerService>& update_policy_controller) {
```
Updated in b4c24f1f707c06aa35646e3881e0fbd9d5502e9b

lordgamez commented on extensions/http-curl/tests/HTTPHandlers.h:
```diff
@@ -446,13 +453,122 @@ class HeartbeatHandler : public ServerAwareHandler {
+        std::vector<std::unordered_map<std::string, std::string>> config_properties;
+        for (const auto& property : minifi::Configuration::CONFIGURATION_PROPERTIES) {
+          std::unordered_map<std::string, std::string> config_property;
+          if (std::find(disallowed_properties.begin(), disallowed_properties.end(), property.name) == disallowed_properties.end()) {
```
Updated in b4c24f1f707c06aa35646e3881e0fbd9d5502e9b

lordgamez commented on extensions/http-curl/tests/HTTPHandlers.h:
```diff
@@ -446,13 +453,122 @@ class HeartbeatHandler : public ServerAwareHandler {
+ private:
+  using Metadata = std::unordered_map<std::string, std::vector<std::unordered_map<std::string, std::string>>>;
+
+  void sendStopOperation(struct mg_connection *conn) {
```
Updated in 99c0658a641e62508440320de372001854efa4c5
